src/ray/raylet/worker_killing_policy_group_by_owner.cc

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/worker_killing_policy_group_by_owner.h"
+
+#include <gtest/gtest_prod.h>
+
+#include <boost/container_hash/hash.hpp>
+#include <unordered_map>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/time/time.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/raylet/worker.h"
+#include "ray/raylet/worker_killing_policy.h"
+#include "ray/raylet/worker_pool.h"
+
+namespace ray {
+
+namespace raylet {
+
+GroupByOwnerIdWorkerKillingPolicy::GroupByOwnerIdWorkerKillingPolicy() {}
+
+const std::pair<std::shared_ptr<WorkerInterface>, bool>
+GroupByOwnerIdWorkerKillingPolicy::SelectWorkerToKill(
+    const std::vector<std::shared_ptr<WorkerInterface>> &workers,
+    const MemorySnapshot &system_memory) const {
+  if (workers.empty()) {
+    RAY_LOG_EVERY_MS(INFO, 5000) << "Worker list is empty. Nothing can be killed";
+    return std::make_pair(nullptr, /*should retry*/ false);
+  }
+
+  TaskID non_retriable_owner_id = TaskID::Nil();
+  std::unordered_map<TaskID, Group> group_map;
+  for (auto worker : workers) {
+    bool retriable = worker->GetAssignedTask().GetTaskSpecification().IsRetriable();
+    if (non_retriable_owner_id == TaskID::Nil()) {
+      non_retriable_owner_id = TaskID::FromRandom(worker->GetAssignedJobId());
```

Comment:
Looks like the value is non-unique, and causing the tests to sporadically fail. I've added a TODO

Reply:
Hmm can you show the test that failed? This would be really concerning if it's true, since we rely on Nil being the same value pretty broadly in the codebase.

Reply:
Hmm looks like i can't repro the failure from before, let me change it to just Nil